catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectExporter.kt

```diff
@@ -0,0 +1,103 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.catroid.io.asynctask
+
+import android.content.Context
+import android.util.Log
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import org.catrobat.catroid.R
+import org.catrobat.catroid.common.Constants
+import org.catrobat.catroid.io.StorageOperations
+import org.catrobat.catroid.io.ZipArchiver
+import org.catrobat.catroid.utils.notifications.NotificationData
+import org.catrobat.catroid.utils.notifications.StatusBarNotificationManager
+import java.io.File
+import java.io.IOException
+import java.lang.ref.WeakReference
+
+class ProjectExporter(
+    private val projectDir: File,
+    private val notificationData: NotificationData,
+    context: Context
+) {
+    private val contextWeakReference: WeakReference<Context> = WeakReference(context)
+
+    @JvmOverloads
+    fun exportProjectToExternalStorageAsync(
+        scope: CoroutineScope = CoroutineScope(Dispatchers.IO)
+    ) {
+        scope.launch {
+            exportProjectToExternalStorage()
+        }
+    }
+
+    fun exportProjectToExternalStorage() {
+        deleteUndoFile(projectDir)
+        val projectZip = File(
+            Constants.EXTERNAL_STORAGE_ROOT_EXPORT_DIRECTORY,
+            projectDir.name + Constants.CATROBAT_EXTENSION
+        )
+        val context = contextWeakReference.get() ?: return
+        Constants.EXTERNAL_STORAGE_ROOT_EXPORT_DIRECTORY.mkdirs()
+        if (!Constants.EXTERNAL_STORAGE_ROOT_EXPORT_DIRECTORY.isDirectory) {
+            return
+        }
+        if (projectZip.exists()) {
+            projectZip.delete()
+        }
+        try {
+            ZipArchiver().zip(projectZip, projectDir.listFiles())
+            StatusBarNotificationManager(context).showOrUpdateNotification(
+                context,
+                notificationData,
+                PROGRESS_IN_PERCENT,
+                null
+            )
+        } catch (e: IOException) {
+            Log.e(TAG, "Cannot create archive.", e)
+            StatusBarNotificationManager(context).abortProgressNotificationWithMessage(
+                context,
+                notificationData,
+                R.string.save_project_to_external_storage_io_exception_message
+            )
+        }
+    }
+
+    private fun deleteUndoFile(projectDir: File) {
+        val undoCodeFile = File(projectDir, Constants.UNDO_CODE_XML_FILE_NAME)
+        if (undoCodeFile.exists()) {
+            try {
+                StorageOperations.deleteFile(undoCodeFile)
+            } catch (exception: IOException) {
+                Log.e(TAG, "Deleting undo file failed.", exception)
+            }
+        }
+    }
+
+    companion object {
+        private val TAG = ProjectExporter::class.java.simpleName
+        private const val PROGRESS_IN_PERCENT = 100
```

Comment:
Please, make it a compile-time constant.
```suggestion
        private const val PROGRESS_FINISHED_PERCENT = 100
```
https://kotlinlang.org/docs/properties.html#compile-time-constants